### ISSUE TITLE
Database is not Migrated from V12 to V14

### ISSUE THREAD

**Issue by EngelenH** [sergeyzwezdin/Hangfire.Mongo#164]
Trying the latest version which replaces a number of collections like statedata with the unified jobGraph collection but despite having the migration options set to migrate and backupcollections no actual migration is done.

In trying to figure out why I am also confused as to the absence of the prefix.schema collection even when making a new repo.

**Comment by EngelenH** [comment]
should point out this was version 0.5.12 and I just noticed there is a 0.5.13 ... Testing to see if that fixes it.

Problem is I can run the upgrade from my dev box but deploying a new version of the application with 0.5.13 would take a bit of effort. Can I run the update with a small dummy app on my devbox using 0.5.13 but once the migration is done use my application with 0.5.12 until I can re-deploy?

**Comment by gottscj** [comment]
if there isnt a prefix.schema, we will assume an empty collection and run all migration steps and in the end add a schema to the db. there should be a schema collection. always.

Im not sure I understand the question correctly, but you should be able run 0.5.12 with schema14 the only migrations in schema14 is minor clean up no breaking changes in the db.

**Comment by EngelenH** [comment]
Well I just ran the 0.5.13 version and this is the first time ever that the .schema collection has appeared for me. Problem is it started from 0.4 and got stuck during upgrade to 0.6 as the schema I am currently using is actually much newer (version 12). I changed the _id in the 1 document in .schema to 12 but now run into another issue which is a key too large error during the creation of an index.

I theory I should be able to circumvent that with running mongod like this : 
mongod --setParameter failIndexKeyTooLong=false

**Comment by gottscj** [comment]
stuck? would you be able to provide some more info? im also curious how you were able to run hangfire without the schema collection. :)

would you be able to drop your jobs db? do you have any unfinished jobs? 

else you could clone the repo and use/hack the migration manager to run the needed migrations? on a backup of course?

**Comment by gottscj** [comment]
It could also be interesting to see which index is problematic. shouldnt need to do special configuration of mongodb

**Comment by EngelenH** [comment]
It's always run like this for months now. Dropping the DB is a bit of a no-go if I can avoid it as it has history in it that is kind of needed. Hacking the migration manager is pretty much what I am doing atm :)

As for stuck well the upgrade stops during the attempted creation of the index, I think this is probably due to me using Hangfire.Console too which logs a lot of the job output in the same DB but this output is potentially quite long. Mongo is limited with regards to how long an index key is allowed to be out of the box.

**Comment by EngelenH** [comment]
Well had to start from scratch in the end anyways. Even adding the failIndexKeyTooLong parameter was not enough (It did fix the index issue).

I think the migration mechanism probably could use some work. It seems to me it is far too strict. Crap happens when you run any app for a year or more. Once the index issue was solved the migration process started to recombine all the old data into the single collection and quickly found missing data for an old job.

Now this is probably not uncommon, software crashes, servers go down unexpectedly etc. Imho the migration mechanism should have an option to do a 'best effort' migration. One where errors can be ignored even if it leads to some data loss.

Now it just gives an error and stops migrating essentially making migration impossible.

**Comment by EngelenH** [comment]
Bit problematic, financial institutions are picky about those things for some odd reason :) 

mostly along these lines though :
ERROR 10:29:00 Error occurred during execution of 'RecurringJobScheduler' process. Execution will be retried (attempt #768) in 00:05:00 seconds.

System.FormatException: Element 'Field' does not match any field or property of class Hangfire.Mongo.Dto.HashDto.

**Comment by gottscj** [comment]
@EngelenH,

Makes sense. 

All those objects are deleted in migration step towards schema14. you can maunally delete all dto's using this command in the mongo shell:
db['hangfire.jobGraph'].deleteMany({ "_t.3": "HashDto", $and: [ { "Field": { $exists: true } } ] })

I introduced a bug in schema13 when aggregating data. HashDto objects now have an embedded document called "Fields" instead of multiple HashDto objects with with the same key, but one "Field" each.

You could check that you have a HashDto with a embedded document called "Fields" and multiple with same key with a field called "Field", if thats the case you can delete all the documents with the field "Field" as the mongo command does above.

You can see the actual migration step here:
https://github.com/sergeyzwezdin/Hangfire.Mongo/blob/master/src/Hangfire.Mongo/Migration/Steps/Version14/00_RemoveMergedHashDtos.cs
